49.md

```diff
@@ -0,0 +1,100 @@
+having the user copy-paste a wallet connection URI into the app they wish to connect with. This can be a UX hurdle and
+often has the user handling sensitive information that they may not understand.
+
+As well, Nostr Wallet Connect has lots of different functions that a wallet may or may not support or an app may or may
+not need. This either runs into silent failures or requires the user to understand what functions their wallet supports
+and what functions the app needs.
+
+Finally, Nostr Wallet Connect requires the wallet to generate both key pairs and the app to use these. This limits the
+key management strategies that an app can use and often leads to suboptimal key management.
+
+This NIP proposes a new protocol that solves these problems by having the wallet and app generate a NWC connection URI
+together. This URI is then used to connect the wallet and app.
+
+## Terms
+
+- **app**: Nostr app on any platform that wants to permissioned access to a user's lightning wallet.
+- **wallet**: A nostr wallet connect capable lightning wallet that the user wants to connect to the app.
+
+## Theory of Operation
+
+1. The user wants to connect their wallet to an app. The app provides a QR code or other means of connecting to the app.
+   The QR code contains a `nostr+walletauth://` URI that contains the app's public key, the relay to communicate over,
+   and a list of functions that the app requires.
+2. The user scans the QR code and their wallet prompts them the required functions and asks them to confirm that they
+   want to connect to the app.
+3. On confirmation, the wallet generates a new key pair and sends the public key to the app in a confirmation event.
+4. With the combination of both keys, the wallet and the app both will generate the same connection URI. They can now
+   use NWC to communicate.
+
+## Specification
+
+### URI Format
+
+The **app** generates this connection URI with protocol `nostr+walletauth:` and base path it's hex-encoded `pubkey`
+with the following query string parameters:
+
+- `relay` Required. URL of the relay where the **app** is connected and will be listening for events. May be
+  more than one.
+- `secret` Required. A random identifier that the **wallet** will use to identify the connection. Without this the
+  **app** cannot authenticate they are receiving a message from the user they want, it could be an attacker just sending
+  kind `33194` events to the **app**. This should be unique for each connection.
+- `required_commands` Required. A space-separated list of commands that the **wallet** requires. The **wallet**
+  MUST NOT connect if it does not support all of these permissions.
+- `optional_commands` Optional. A space-separated list of commands that the **wallet** can enable to add additional
+  functionality. The **wallet** MAY ignore these.
+- `budget` Optional. A budget that the **wallet** will use to limit the amount of funds that the **app** can spend.
```

Comment:
I think having a budget is great, but should it really be up to the **app** to decide this budget? On one hand, it's nice for the app to be able to self-limit (I promise I'll only send 10 sats per day), but I think it might also be nice to give the wallet and even the user the ability to change this budget, like Alby allows today.

Reply:
This is just a recommendation from the app, it is not required by the wallet and can be changed. You can see how we do this with mutiny + zapple pay if you're curious